### PR TITLE
Globals (GHC boot packages) pruning

### LINKED ISSUE #4510: Rebuilding packages from the global database with modified dependencies
This applies to Stack master, specifically commit 9900d99114967698223f43c84e4cedecc24cea97. Repro (on a non-Windows system). With the following `stack.yaml`:

```yaml
resolver: ghc-8.6.3
packages: []
extra-deps:
- time-1.8.0.1
```

This replaces the global `time` package with a different version. Running `stack build unix` causes both `time` and `unix` to be compiled, which makes sense, since `unix` depends on `time`.

Running `stack build unix` a second time should be a no-op, but instead, it rebuilds:

```
$ stack build unix
unix-2.7.2.2: configure
^C
```

This appears to be because the `SourceMap` knows nothing about `unix` (since it's a global package and not present at all in snapshot), and therefore Stack is ignoring the already-present copy in the snapshot database.

Possible solution: update the SourceMap to contain information about global packages.

CC @qrilka

### PR DESCRIPTION
**Pull request by qrilka**
Please include the following checklist in your PR:

* [ ] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [x] The documentation has been updated, if necessary.

Tested manually, plan to run integration tests.
Failure to build global packages as targets results in 2 errors (using `stack.yaml` from #4510 with `time` overriden):
```
$ stack build directory --dry-run

Error: While constructing the build plan, the following exceptions were encountered:

Can't use GHC boot package directory when it has an overriden dependency, you need to add the following as explicit dependencies to the
project: 
 unix

Some different approaches to resolving this:

  * Consider trying 'stack solver', which uses the cabal-install solver to attempt to find some working build configuration. This can be
    convenient when dealing with many complicated constraint errors, but results may be unpredictable.


Plan construction failed.
```
or
```
$ stack build process --dry-run

Error: While constructing the build plan, the following exceptions were encountered:

In the dependencies for process-1.6.5.0:
    directory must match >=1.1 && <1.4, but the stack configuration has no specified version  (latest matching version is 1.3.3.2)
    unix must match >=2.5 && <2.9, but the stack configuration has no specified version  (latest matching version is 2.7.2.2)
needed since process is a build target.

Some different approaches to resolving this:

  * Consider trying 'stack solver', which uses the cabal-install solver to attempt to find some working build configuration. This can be
    convenient when dealing with many complicated constraint errors, but results may be unpredictable.

  * Recommended action: try adding the following to your extra-deps in /home/qrilka/ws/h/stack4510/stack.yaml:

directory-1.3.3.2@sha256:79e15a62f90cb9f20f46b2b56fe67d00d71ed54cfad07ea337695f9ebb74baa4,2829
unix-2.7.2.2@sha256:4ef1d010d70a4a07a717e853d4a440c105dad38c6199316e320fdd4c48dacd34,3495

Plan construction failed.
```
Fixes #4510

Also I plan to update curator to take these details into account (probably as a separate PR)

### CODE REVIEW

File: src/Stack/SourceMap.hs [hunk 1]
```diff
@@ -111,26 +113,26 @@ getPLIVersion (PLIRepo _ pm) = pkgVersion $ pmIdent pm
...
             ( pkgName $ dpPackageIdent d
-            , GlobalPackage (pkgVersion $ dpPackageIdent d))
+            , pkgVersion $ dpPackageIdent d)
     toGlobals <$> ghcPkgDump (whichCompiler compiler) [] pkgConduit
 
 globalsFromHints ::
        HasConfig env
     => WantedCompiler
-    -> RIO env (Map PackageName GlobalPackage)
+    -> RIO env (Map PackageName Version)
```
Comment (snoyberg) [inline src/Stack/SourceMap.hs]: I'm almost tempted to say we should have a special `GlobalVersion` newtype wrapper to drive the point home further, but this is good.

**Comment by qrilka** [comment]
I have fixed integration test `345-override-bytestring` and added this pruning to curator. It looks like `build-constraints.yaml` doesn't include `parsec` (while `mtl` is overriden) - PR on https://github.com/commercialhaskell/stackage will follow

**Comment by qrilka** [comment]
This package pruning implementation partially relies on information in `global-hints.yaml` but as e.g. Michael demonstrated in his blog post packages shipped with GHC could have no (or bad) corresponding version on Hackage so I'm changing it to to use correct package information from `ghc-pkg`

**Comment by qrilka** [comment]
I have changed the code quite a lot, also it looks like the refactoring I promised almost a month ago https://github.com/commercialhaskell/stack/issues/4480#issuecomment-452037476 could be quite helpful in improving code around `SkipCompilerDownload`.
@snoyberg could I ask you for another look?

**Pull request merged**